pkg/codegen/hcl2/model/type.go

```diff
@@ -43,13 +45,84 @@ type Type interface {
 	ConversionFrom(src Type) ConversionKind
 	String() string
 
+	hash(stack objTypeSet) uint32
 	equals(other Type, seen map[Type]struct{}) bool
 	conversionFrom(src Type, unifying bool, seen map[Type]struct{}) (ConversionKind, lazyDiagnostics)
 	string(seen map[Type]struct{}) string
 	unify(other Type) (Type, ConversionKind)
 	isType()
 }
 
+type objTypeSet map[*ObjectType]struct{}
+
+func (s *objTypeSet) add(t *ObjectType) bool {
+	m := *s
+	if m == nil {
+		m = map[*ObjectType]struct{}{}
+		*s = m
+	}
+
+	_, has := m[t]
+	m[t] = struct{}{}
+	return has
+}
+
+func (s objTypeSet) delete(t *ObjectType) {
+	delete(s, t)
+}
+
+const (
+	hashKindNone uint32 = iota + 1
+	hashKindConst
+	hashKindList
+	hashKindMap
+	hashKindNamed
+	hashKindObject
+	hashKindOpaque
+	hashKindOutput
+	hashKindPromise
+	hashKindSet
+	hashKindTuple
+	hashKindUnion
+)
+
+//go:linkname hashString runtime.stringHash
+func hashString(s string, seed uintptr) uintptr
+
+type fnvHash uint32
```

Comment:
Considered linking in fnv1a from https://github.com/segmentio/fasthash ? Don't know how conservative we are about taking on little dependencies. If you think we prefer rolling our own a quick comment on why in here would be good.

Reply:
> If you think we prefer rolling our own a quick comment on why in here would be good.

That's a good find. I rolled this one together because
a. FNV-1a is simple
b. The [stdlib implementation](https://pkg.go.dev/hash/fnv@go1.16.6#New32a) doesn't have a convenience method for adding `Uint32` values to the hash, and writing a full FNV-1a implementation is ~as much code as writing a convenience method
c. The ability to use the internal implementation of string hashing is useful for performance.

All that said, I think that we can probably use the package you've linked w/o issue. Really wish the stdlib had the convenience method I wanted...